src/hugr/typecheck.rs

```diff
@@ -0,0 +1,163 @@
+//! Simple type checking - takes a hugr and some extra info and checks whether
+//! the types at the sources of each wire match those of the targets
+
+use crate::hugr::*;
+use crate::types::{SimpleType, TypeRow};
+
+// For static typechecking
+use crate::ops::ConstValue;
+use crate::types::{ClassicType, Container};
+
+use std::fmt::{self, Display};
+
+/// Errors that arise from typechecking constants
+#[derive(Clone, Debug, Eq, PartialEq, Error)]
+pub enum TypeError {
+    /// This case hasn't been implemented. Possibly because we don't have value
+    /// constructors to check against it
+    Unimplemented(ClassicType),
+    /// The given type and term are incompatible
+    Failed(ClassicType),
+    /// The value exceeds the max value of its `I<n>` type
+    /// E.g. checking 300 against I8
+    IntTooLarge(u8, u128),
+    /// Width (n) of an `I<n>` type doesn't fit into a u32
+    IntWidthTooLarge(u8),
+    /// The width of an integer type wasn't a power of 2
+    IntWidthInvalid(u8),
+    /// Expected width (packed with const int) doesn't match type
+    IntWidthMismatch(u8, u8),
+    /// Found a Var type constructor when we're checking a const val
+    ConstCantBeVar,
+    /// The length of the tuple value doesn't match the length of the tuple type
+    TupleWrongLength,
+    /// Const values aren't allowed to be linear
+    LinearTypeDisallowed,
+    /// Tag for a sum value exceeded the number of variants
+    InvalidSumTag,
+    /// For a value which embeds its type (e.g. sum or opaque) - a mismatch
+    /// between the embedded type and the type we're checking against
+    TypeMismatch(ClassicType, ClassicType),
+    /// A mismatch between the embedded type and the type we're checking
+    /// against, as above, but for rows instead of simple types
+    TypeRowMismatch(TypeRow, TypeRow),
+}
+
+impl Display for TypeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let str = match self {
+            TypeError::Unimplemented(ty) => {
+                format!("Const type checking unimplemented for {}", ty)
```

Comment:
`thiserror` makes this much easier,

see for example https://github.com/CQCL-DEV/hugr/blob/f80dce7105c4d99dec904d6ef33930a584d3dc9a/src/hugr.rs#L238